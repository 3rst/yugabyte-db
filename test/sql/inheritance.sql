\unset ECHO
\i test/setup.sql
SELECT plan( 85 );
--SELECT * FROM no_plan();
SET client_min_messages = warning;

-- Create inherited tables
CREATE TABLE public.parent( id INT PRIMARY KEY );
CREATE TABLE public.child1( id INT PRIMARY KEY ) INHERITS ( public.parent );
CREATE TABLE public.child2( id INT PRIMARY KEY ) INHERITS ( public.child1 );

-- Create inherited tables in another schema
CREATE SCHEMA hide;
CREATE TABLE hide.h_parent( id INT PRIMARY KEY );
CREATE TABLE hide.h_child1( id INT PRIMARY KEY ) INHERITS ( hide.h_parent );
CREATE TABLE hide.h_child2( id INT PRIMARY KEY ) INHERITS ( hide.h_child1 );

-- test has_inhereted_tables
SELECT * FROM check_test(
    has_inherited_tables( 'hide', 'h_parent', 'Gimme inheritance' ),
    true,
    'has_inherited_tables(sch, tab, desc)',
    'Gimme inheritance',
    ''
);

SELECT * FROM check_test(
    has_inherited_tables( 'hide', 'h_child2', 'Gimme inheritance' ),
    false,
    'has_inherited_tables(sch, tab, desc) fail',
    'Gimme inheritance',
    ''
);

SELECT * FROM check_test(
    has_inherited_tables( 'hide', 'nonesuch', 'Gimme inheritance' ),
    false,
    'has_inherited_tables(sch, nonesuch, desc)',
    'Gimme inheritance',
    ''
);

SELECT * FROM check_test(
    has_inherited_tables( 'hide', 'h_parent'::name ),
    true,
    'has_inherited_tables(sch, tab)',
    'Table hide.h_parent should have children',
    ''
);

SELECT * FROM check_test(
    has_inherited_tables( 'hide', 'h_child2'::name ),
    false,
    'has_inherited_tables(sch, tab) fail',
    'Table hide.h_child2 should have children',
    ''
);

SELECT * FROM check_test(
    has_inherited_tables( 'hide', 'nonesuch'::name ),
    false,
    'has_inherited_tables(sch, nonesuch)',
    'Table hide.nonesuch should have children',
    ''
);

SELECT * FROM check_test(
    has_inherited_tables( 'parent', 'Gimme more' ),
    true,
    'has_inherited_tables(tab, desc)',
    'Gimme more',
    ''
);

SELECT * FROM check_test(
    has_inherited_tables( 'child2', 'Gimme more' ),
    false,
    'has_inherited_tables(tab, desc) fail',
    'Gimme more',
    ''
);

SELECT * FROM check_test(
    has_inherited_tables( 'nonesuch', 'Gimme more' ),
    false,
    'has_inherited_tables(nonesuch, desc)',
    'Gimme more',
    ''
);

SELECT * FROM check_test(
    has_inherited_tables( 'parent' ),
    true,
    'has_inherited_tables(tab)',
    'Table parent should have children',
    ''
);

SELECT * FROM check_test(
    has_inherited_tables( 'child2' ),
    false,
    'has_inherited_tables(tab) fail',
    'Table child2 should have children',
    ''
);

SELECT * FROM check_test(
    has_inherited_tables( 'nonesuch' ),
    false,
    'has_inherited_tables(nonesuch)',
    'Table nonesuch should have children',
    ''
);

-- test hasnt_inherited_tables
SELECT * FROM check_test(
    hasnt_inherited_tables( 'hide', 'h_child2', 'Gimme inheritance' ),
    true,
    'hasnt_inherited_tables(sch, tab, desc)',
    'Gimme inheritance',
    ''
);

SELECT * FROM check_test(
    hasnt_inherited_tables( 'hide', 'h_child1', 'Gimme inheritance' ),
    false,
    'hasnt_inherited_tables(sch, tab, desc) fail',
    'Gimme inheritance',
    ''
);

SELECT * FROM check_test(
    hasnt_inherited_tables( 'hide', 'nonesuch', 'Gimme inheritance' ),
    true,
    'hasnt_inherited_tables(sch, nonesuch, desc)',
    'Gimme inheritance',
    ''
);

SELECT * FROM check_test(
    hasnt_inherited_tables( 'hide', 'h_child2'::name ),
    true,
    'hasnt_inherited_tables(sch, tab)',
    'Table hide.h_child2 should not have children',
    ''
);

SELECT * FROM check_test(
    hasnt_inherited_tables( 'hide', 'h_child1'::name ),
    false,
    'hasnt_inherited_tables(sch, tab) fail',
    'Table hide.h_child1 should not have children',
    ''
);

SELECT * FROM check_test(
    hasnt_inherited_tables( 'hide', 'nonesuch'::name ),
    true,
    'hasnt_inherited_tables(sch, nonesuch)',
    'Table hide.nonesuch should not have children',
    ''
);

SELECT * FROM check_test(
    hasnt_inherited_tables( 'child2', 'Gimme inheritance' ),
    true,
    'hasnt_inherited_tables(tab, desc)',
    'Gimme inheritance',
    ''
);

SELECT * FROM check_test(
    hasnt_inherited_tables( 'child1', 'Gimme inheritance' ),
    false,
    'hasnt_inherited_tables(tab, desc) fail',
    'Gimme inheritance',
    ''
);

SELECT * FROM check_test(
    hasnt_inherited_tables( 'nonesuch', 'Gimme inheritance' ),
    true,
    'hasnt_inherited_tables(nonesuch, desc)',
    'Gimme inheritance',
    ''
);

SELECT * FROM check_test(
    hasnt_inherited_tables( 'child2' ),
    true,
    'hasnt_inherited_tables(tab)',
    'Table child2 should not have children',
    ''
);

SELECT * FROM check_test(
    hasnt_inherited_tables( 'child1' ),
    false,
    'hasnt_inherited_tables(tab) fail',
    'Table child1 should not have children',
    ''
);

SELECT * FROM check_test(
    hasnt_inherited_tables( 'nonesuch' ),
    true,
    'hasnt_inherited_tables(nonesuch)',
    'Table nonesuch should not have children',
    ''
);





SELECT * FROM check_test(
    is_parent_of( 'hide', 'h_parent', 'hide', 'h_child1', 1, 'Test hide.h_parent->hide.h_child1' ),
    true, -- expected value
    'hide.h_parent direct is father of hide.h_child1'
);

SELECT * FROM check_test(
    is_parent_of( 'hide', 'h_parent', 'hide', 'h_child1', 1 ),
    true, -- expected value
    'hide.h_parent direct is father of hide.h_child1'
);

SELECT * FROM check_test(
    is_parent_of( 'hide', 'h_child1', 'hide', 'h_child2', 1 ),
    true, -- expected value
    'hide.h_child1 direct is father of hide.h_child2'
);

SELECT * FROM check_test(
    is_parent_of( 'hide', 'h_parent', 'hide', 'h_child2', 2 ),
    true, -- expected value
    'hide.h_parent is father of hide.h_child2'
);

SELECT * FROM check_test(
    is_parent_of( 'parent', 'child1' ),
    true, -- expected value
    'child1 inherits from parent'
);

SELECT * FROM check_test(
    is_parent_of( 'hide'::name, 'h_parent'::name, 'public'::name, 'child1'::name ),
    false, -- expected value
    'hide.h_parent is not father of public.child1'
);

SELECT * FROM check_test(
    is_parent_of( 'public'::name, 'parent'::name, 'public'::name, 'child1'::name ),
    true, -- expected value
    'public.parent is not father of public.child1'
);

SELECT * FROM check_test(
    isnt_child_of( 'hide'::name, 'h_parent'::name, 'public'::name, 'child1'::name ),
    true, -- expected value
    'hide.h_parent is not father of public.child1'
);

SELECT * FROM check_test(
    isnt_parent_of( 'parent', 'child1' ),
    false, -- expected value
    'parent is not father of public.child1'
);

SELECT * FROM check_test(
    isnt_child_of( 'parent', 'child1' ),
    true, -- expected value
    'parent is not child1'
);

SELECT * FROM check_test(
    isnt_child_of( 'child1', 'parent' ),
    false, -- expected value
    'child1 inherits from parent'
);

SELECT * FROM check_test(
    isnt_child_of( 'hide'::name, 'h_parent'::name, 'hide'::name, 'h_child1'::name ),
    true, -- expected value
    'hide.h_parent is not child of hide.h_child1'
);

SELECT * FROM check_test(
    isnt_child_of( 'hide'::name, 'h_child1'::name, 'hide'::name, 'h_parent'::name ),
    false, -- expected value
    'hide.h_child1 inherits from hide.h_parent'
);

/****************************************************************************/
-- Finish the tests and clean up.
SELECT * FROM finish();
ROLLBACK;
