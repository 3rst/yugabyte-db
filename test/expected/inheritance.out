\unset ECHO
1..240
ok 1 - has_inherited_tables(sch, tab, desc) should pass
ok 2 - has_inherited_tables(sch, tab, desc) should have the proper description
ok 3 - has_inherited_tables(sch, tab, desc) should have the proper diagnostics
ok 4 - has_inherited_tables(sch, tab, desc) fail should fail
ok 5 - has_inherited_tables(sch, tab, desc) fail should have the proper description
ok 6 - has_inherited_tables(sch, tab, desc) fail should have the proper diagnostics
ok 7 - has_inherited_tables(sch, nonesuch, desc) should fail
ok 8 - has_inherited_tables(sch, nonesuch, desc) should have the proper description
ok 9 - has_inherited_tables(sch, nonesuch, desc) should have the proper diagnostics
ok 10 - has_inherited_tables(sch, tab) should pass
ok 11 - has_inherited_tables(sch, tab) should have the proper description
ok 12 - has_inherited_tables(sch, tab) should have the proper diagnostics
ok 13 - has_inherited_tables(sch, tab) fail should fail
ok 14 - has_inherited_tables(sch, tab) fail should have the proper description
ok 15 - has_inherited_tables(sch, tab) fail should have the proper diagnostics
ok 16 - has_inherited_tables(sch, nonesuch) should fail
ok 17 - has_inherited_tables(sch, nonesuch) should have the proper description
ok 18 - has_inherited_tables(sch, nonesuch) should have the proper diagnostics
ok 19 - has_inherited_tables(tab, desc) should pass
ok 20 - has_inherited_tables(tab, desc) should have the proper description
ok 21 - has_inherited_tables(tab, desc) should have the proper diagnostics
ok 22 - has_inherited_tables(tab, desc) fail should fail
ok 23 - has_inherited_tables(tab, desc) fail should have the proper description
ok 24 - has_inherited_tables(tab, desc) fail should have the proper diagnostics
ok 25 - has_inherited_tables(nonesuch, desc) should fail
ok 26 - has_inherited_tables(nonesuch, desc) should have the proper description
ok 27 - has_inherited_tables(nonesuch, desc) should have the proper diagnostics
ok 28 - has_inherited_tables(tab) should pass
ok 29 - has_inherited_tables(tab) should have the proper description
ok 30 - has_inherited_tables(tab) should have the proper diagnostics
ok 31 - has_inherited_tables(tab) fail should fail
ok 32 - has_inherited_tables(tab) fail should have the proper description
ok 33 - has_inherited_tables(tab) fail should have the proper diagnostics
ok 34 - has_inherited_tables(nonesuch) should fail
ok 35 - has_inherited_tables(nonesuch) should have the proper description
ok 36 - has_inherited_tables(nonesuch) should have the proper diagnostics
ok 37 - hasnt_inherited_tables(sch, tab, desc) should pass
ok 38 - hasnt_inherited_tables(sch, tab, desc) should have the proper description
ok 39 - hasnt_inherited_tables(sch, tab, desc) should have the proper diagnostics
ok 40 - hasnt_inherited_tables(sch, tab, desc) fail should fail
ok 41 - hasnt_inherited_tables(sch, tab, desc) fail should have the proper description
ok 42 - hasnt_inherited_tables(sch, tab, desc) fail should have the proper diagnostics
ok 43 - hasnt_inherited_tables(sch, nonesuch, desc) should pass
ok 44 - hasnt_inherited_tables(sch, nonesuch, desc) should have the proper description
ok 45 - hasnt_inherited_tables(sch, nonesuch, desc) should have the proper diagnostics
ok 46 - hasnt_inherited_tables(sch, tab) should pass
ok 47 - hasnt_inherited_tables(sch, tab) should have the proper description
ok 48 - hasnt_inherited_tables(sch, tab) should have the proper diagnostics
ok 49 - hasnt_inherited_tables(sch, tab) fail should fail
ok 50 - hasnt_inherited_tables(sch, tab) fail should have the proper description
ok 51 - hasnt_inherited_tables(sch, tab) fail should have the proper diagnostics
ok 52 - hasnt_inherited_tables(sch, nonesuch) should pass
ok 53 - hasnt_inherited_tables(sch, nonesuch) should have the proper description
ok 54 - hasnt_inherited_tables(sch, nonesuch) should have the proper diagnostics
ok 55 - hasnt_inherited_tables(tab, desc) should pass
ok 56 - hasnt_inherited_tables(tab, desc) should have the proper description
ok 57 - hasnt_inherited_tables(tab, desc) should have the proper diagnostics
ok 58 - hasnt_inherited_tables(tab, desc) fail should fail
ok 59 - hasnt_inherited_tables(tab, desc) fail should have the proper description
ok 60 - hasnt_inherited_tables(tab, desc) fail should have the proper diagnostics
ok 61 - hasnt_inherited_tables(nonesuch, desc) should pass
ok 62 - hasnt_inherited_tables(nonesuch, desc) should have the proper description
ok 63 - hasnt_inherited_tables(nonesuch, desc) should have the proper diagnostics
ok 64 - hasnt_inherited_tables(tab) should pass
ok 65 - hasnt_inherited_tables(tab) should have the proper description
ok 66 - hasnt_inherited_tables(tab) should have the proper diagnostics
ok 67 - hasnt_inherited_tables(tab) fail should fail
ok 68 - hasnt_inherited_tables(tab) fail should have the proper description
ok 69 - hasnt_inherited_tables(tab) fail should have the proper diagnostics
ok 70 - hasnt_inherited_tables(nonesuch) should pass
ok 71 - hasnt_inherited_tables(nonesuch) should have the proper description
ok 72 - hasnt_inherited_tables(nonesuch) should have the proper diagnostics
ok 73 - is_ancestor_of(psch, ptab, csch, ctab, 1, desc) should pass
ok 74 - is_ancestor_of(psch, ptab, csch, ctab, 1, desc) should have the proper description
ok 75 - is_ancestor_of(psch, ptab, csch, ctab, 1, desc) should have the proper diagnostics
ok 76 - is_ancestor_of(psch, ptab, csch, ctab, 2, desc) should pass
ok 77 - is_ancestor_of(psch, ptab, csch, ctab, 2, desc) should have the proper description
ok 78 - is_ancestor_of(psch, ptab, csch, ctab, 2, desc) should have the proper diagnostics
ok 79 - is_ancestor_of(psch, nope, csch, ctab, 1, desc) should fail
ok 80 - is_ancestor_of(psch, nope, csch, ctab, 1, desc) should have the proper description
ok 81 - is_ancestor_of(psch, nope, csch, ctab, 1, desc) should have the proper diagnostics
ok 82 - is_ancestor_of(psch, ptab, csch, nope, desc) should fail
ok 83 - is_ancestor_of(psch, ptab, csch, nope, desc) should have the proper description
ok 84 - is_ancestor_of(psch, ptab, csch, nope, desc) should have the proper diagnostics
ok 85 - is_ancestor_of(psch, ptab, csch, ctab, 1) should pass
ok 86 - is_ancestor_of(psch, ptab, csch, ctab, 1) should have the proper description
ok 87 - is_ancestor_of(psch, ptab, csch, ctab, 1) should have the proper diagnostics
ok 88 - is_ancestor_of(psch, nope, csch, ctab, 1) should fail
ok 89 - is_ancestor_of(psch, nope, csch, ctab, 1) should have the proper description
ok 90 - is_ancestor_of(psch, nope, csch, ctab, 1) should have the proper diagnostics
ok 91 - is_ancestor_of(psch, ptab, csch, nope, 1) should fail
ok 92 - is_ancestor_of(psch, ptab, csch, nope, 1) should have the proper description
ok 93 - is_ancestor_of(psch, ptab, csch, nope, 1) should have the proper diagnostics
ok 94 - is_ancestor_of(psch, ptab, csch, ctab, 2) should pass
ok 95 - is_ancestor_of(psch, ptab, csch, ctab, 2) should have the proper description
ok 96 - is_ancestor_of(psch, ptab, csch, ctab, 2) should have the proper diagnostics
ok 97 - is_ancestor_of(psch, nope, csch, ctab, 2) should fail
ok 98 - is_ancestor_of(psch, nope, csch, ctab, 2) should have the proper description
ok 99 - is_ancestor_of(psch, nope, csch, ctab, 2) should have the proper diagnostics
ok 100 - is_ancestor_of(psch, ptab, csch, nope, 2) should fail
ok 101 - is_ancestor_of(psch, ptab, csch, nope, 2) should have the proper description
ok 102 - is_ancestor_of(psch, ptab, csch, nope, 2) should have the proper diagnostics
ok 103 - is_ancestor_of(psch, ptab, csch, ctab, desc) should pass
ok 104 - is_ancestor_of(psch, ptab, csch, ctab, desc) should have the proper description
ok 105 - is_ancestor_of(psch, ptab, csch, ctab, desc) should have the proper diagnostics
ok 106 - is_ancestor_of(psch, ptab, csch, ctab2, desc) should pass
ok 107 - is_ancestor_of(psch, ptab, csch, ctab2, desc) should have the proper description
ok 108 - is_ancestor_of(psch, ptab, csch, ctab2, desc) should have the proper diagnostics
ok 109 - is_ancestor_of(psch, ptab, csch, ctab) should pass
ok 110 - is_ancestor_of(psch, ptab, csch, ctab) should have the proper description
ok 111 - is_ancestor_of(psch, ptab, csch, ctab) should have the proper diagnostics
ok 112 - is_ancestor_of(psch, ptab, csch, ctab2) should pass
ok 113 - is_ancestor_of(psch, ptab, csch, ctab2) should have the proper description
ok 114 - is_ancestor_of(psch, ptab, csch, ctab2) should have the proper diagnostics
ok 115 - is_ancestor_of(psch, nope, csch, ctab) should fail
ok 116 - is_ancestor_of(psch, nope, csch, ctab) should have the proper description
ok 117 - is_ancestor_of(psch, nope, csch, ctab) should have the proper diagnostics
ok 118 - is_ancestor_of(psch, ptab, csch, nope) should fail
ok 119 - is_ancestor_of(psch, ptab, csch, nope) should have the proper description
ok 120 - is_ancestor_of(psch, ptab, csch, nope) should have the proper diagnostics
ok 121 - is_ancestor_of(ptab, ctab, 1, desc) should pass
ok 122 - is_ancestor_of(ptab, ctab, 1, desc) should have the proper description
ok 123 - is_ancestor_of(ptab, ctab, 1, desc) should have the proper diagnostics
ok 124 - is_ancestor_of(ptab, ctab, 2, desc) should pass
ok 125 - is_ancestor_of(ptab, ctab, 2, desc) should have the proper description
ok 126 - is_ancestor_of(ptab, ctab, 2, desc) should have the proper diagnostics
ok 127 - is_ancestor_of(ptab, ctab, 1, desc) fail should fail
ok 128 - is_ancestor_of(ptab, ctab, 1, desc) fail should have the proper description
ok 129 - is_ancestor_of(ptab, ctab, 1, desc) fail should have the proper diagnostics
ok 130 - is_ancestor_of(ptab, nope, 1, desc) should fail
ok 131 - is_ancestor_of(ptab, nope, 1, desc) should have the proper description
ok 132 - is_ancestor_of(ptab, nope, 1, desc) should have the proper diagnostics
ok 133 - is_ancestor_of(ptab, ctab, 1) should pass
ok 134 - is_ancestor_of(ptab, ctab, 1) should have the proper description
ok 135 - is_ancestor_of(ptab, ctab, 1) should have the proper diagnostics
ok 136 - is_ancestor_of(ptab, ctab, 2) should pass
ok 137 - is_ancestor_of(ptab, ctab, 2) should have the proper description
ok 138 - is_ancestor_of(ptab, ctab, 2) should have the proper diagnostics
ok 139 - is_ancestor_of(ptab, ctab, 1) fail should fail
ok 140 - is_ancestor_of(ptab, ctab, 1) fail should have the proper description
ok 141 - is_ancestor_of(ptab, ctab, 1) fail should have the proper diagnostics
ok 142 - is_ancestor_of(ptab, nope, 1) should fail
ok 143 - is_ancestor_of(ptab, nope, 1) should have the proper description
ok 144 - is_ancestor_of(ptab, nope, 1) should have the proper diagnostics
ok 145 - is_ancestor_of(ptab, ctab) should pass
ok 146 - is_ancestor_of(ptab, ctab) should have the proper description
ok 147 - is_ancestor_of(ptab, ctab) should have the proper diagnostics
ok 148 - is_ancestor_of(ptab, ctab2) should pass
ok 149 - is_ancestor_of(ptab, ctab2) should have the proper description
ok 150 - is_ancestor_of(ptab, ctab2) should have the proper diagnostics
ok 151 - is_ancestor_of(ptab, nope) should fail
ok 152 - is_ancestor_of(ptab, nope) should have the proper description
ok 153 - is_ancestor_of(ptab, nope) should have the proper diagnostics
ok 154 - is_ancestor_of(nope, ctab2) should fail
ok 155 - is_ancestor_of(nope, ctab2) should have the proper description
ok 156 - is_ancestor_of(nope, ctab2) should have the proper diagnostics
ok 157 - is_descendent_of(csch, ctab, psch, ptab, 1, desc) should pass
ok 158 - is_descendent_of(csch, ctab, psch, ptab, 1, desc) should have the proper description
ok 159 - is_descendent_of(csch, ctab, psch, ptab, 1, desc) should have the proper diagnostics
ok 160 - is_descendent_of(csch, ctab, psch, ptab, 2, desc) should pass
ok 161 - is_descendent_of(csch, ctab, psch, ptab, 2, desc) should have the proper description
ok 162 - is_descendent_of(csch, ctab, psch, ptab, 2, desc) should have the proper diagnostics
ok 163 - is_descendent_of(csch, ctab, psch, nope, 1, desc) should fail
ok 164 - is_descendent_of(csch, ctab, psch, nope, 1, desc) should have the proper description
ok 165 - is_descendent_of(csch, ctab, psch, nope, 1, desc) should have the proper diagnostics
ok 166 - is_descendent_of(csch, nope, psch, ptab, desc) should fail
ok 167 - is_descendent_of(csch, nope, psch, ptab, desc) should have the proper description
ok 168 - is_descendent_of(csch, nope, psch, ptab, desc) should have the proper diagnostics
ok 169 - is_descendent_of(csch, ctab, psch, ptab, 1) should pass
ok 170 - is_descendent_of(csch, ctab, psch, ptab, 1) should have the proper description
ok 171 - is_descendent_of(csch, ctab, psch, ptab, 1) should have the proper diagnostics
ok 172 - is_descendent_of(csch, ctab, psch, nope, 1) should fail
ok 173 - is_descendent_of(csch, ctab, psch, nope, 1) should have the proper description
ok 174 - is_descendent_of(csch, ctab, psch, nope, 1) should have the proper diagnostics
ok 175 - is_descendent_of(csch, nope, psch, ptab, 1) should fail
ok 176 - is_descendent_of(csch, nope, psch, ptab, 1) should have the proper description
ok 177 - is_descendent_of(csch, nope, psch, ptab, 1) should have the proper diagnostics
ok 178 - is_descendent_of(csch, ctab, psch, ptab, 2) should pass
ok 179 - is_descendent_of(csch, ctab, psch, ptab, 2) should have the proper description
ok 180 - is_descendent_of(csch, ctab, psch, ptab, 2) should have the proper diagnostics
ok 181 - is_descendent_of(csch, ctab, psch, nope, 2) should fail
ok 182 - is_descendent_of(csch, ctab, psch, nope, 2) should have the proper description
ok 183 - is_descendent_of(csch, ctab, psch, nope, 2) should have the proper diagnostics
ok 184 - is_descendent_of(csch, nope, psch, ptab, 2) should fail
ok 185 - is_descendent_of(csch, nope, psch, ptab, 2) should have the proper description
ok 186 - is_descendent_of(csch, nope, psch, ptab, 2) should have the proper diagnostics
ok 187 - is_descendent_of(csch, ctab, psch, ptab, desc) should pass
ok 188 - is_descendent_of(csch, ctab, psch, ptab, desc) should have the proper description
ok 189 - is_descendent_of(csch, ctab, psch, ptab, desc) should have the proper diagnostics
ok 190 - is_descendent_of(csch, ctab2, psch, ptab, desc) should pass
ok 191 - is_descendent_of(csch, ctab2, psch, ptab, desc) should have the proper description
ok 192 - is_descendent_of(csch, ctab2, psch, ptab, desc) should have the proper diagnostics
ok 193 - is_descendent_of(csch, ctab, psch, ptab) should pass
ok 194 - is_descendent_of(csch, ctab, psch, ptab) should have the proper description
ok 195 - is_descendent_of(csch, ctab, psch, ptab) should have the proper diagnostics
ok 196 - is_descendent_of(csch, ctab2, psch, ptab) should pass
ok 197 - is_descendent_of(csch, ctab2, psch, ptab) should have the proper description
ok 198 - is_descendent_of(csch, ctab2, psch, ptab) should have the proper diagnostics
ok 199 - is_descendent_of(csch, ctab, psch, nope) should fail
ok 200 - is_descendent_of(csch, ctab, psch, nope) should have the proper description
ok 201 - is_descendent_of(csch, ctab, psch, nope) should have the proper diagnostics
ok 202 - is_descendent_of(csch, nope, psch, ptab) should fail
ok 203 - is_descendent_of(csch, nope, psch, ptab) should have the proper description
ok 204 - is_descendent_of(csch, nope, psch, ptab) should have the proper diagnostics
ok 205 - is_descendent_of(ctab, ptab, 1, desc) should pass
ok 206 - is_descendent_of(ctab, ptab, 1, desc) should have the proper description
ok 207 - is_descendent_of(ctab, ptab, 1, desc) should have the proper diagnostics
ok 208 - is_descendent_of(ctab, ptab, 2, desc) should pass
ok 209 - is_descendent_of(ctab, ptab, 2, desc) should have the proper description
ok 210 - is_descendent_of(ctab, ptab, 2, desc) should have the proper diagnostics
ok 211 - is_descendent_of(ctab, ptab, 1, desc) fail should fail
ok 212 - is_descendent_of(ctab, ptab, 1, desc) fail should have the proper description
ok 213 - is_descendent_of(ctab, ptab, 1, desc) fail should have the proper diagnostics
ok 214 - is_descendent_of(nope, ptab, 1, desc) should fail
ok 215 - is_descendent_of(nope, ptab, 1, desc) should have the proper description
ok 216 - is_descendent_of(nope, ptab, 1, desc) should have the proper diagnostics
ok 217 - is_descendent_of(ctab, ptab, 1) should pass
ok 218 - is_descendent_of(ctab, ptab, 1) should have the proper description
ok 219 - is_descendent_of(ctab, ptab, 1) should have the proper diagnostics
ok 220 - is_descendent_of(ctab, ptab, 2) should pass
ok 221 - is_descendent_of(ctab, ptab, 2) should have the proper description
ok 222 - is_descendent_of(ctab, ptab, 2) should have the proper diagnostics
ok 223 - is_descendent_of(ctab, ptab, 1) fail should fail
ok 224 - is_descendent_of(ctab, ptab, 1) fail should have the proper description
ok 225 - is_descendent_of(ctab, ptab, 1) fail should have the proper diagnostics
ok 226 - is_descendent_of(nope, ptab, 1) should fail
ok 227 - is_descendent_of(nope, ptab, 1) should have the proper description
ok 228 - is_descendent_of(nope, ptab, 1) should have the proper diagnostics
ok 229 - is_descendent_of(ctab, ptab) should pass
ok 230 - is_descendent_of(ctab, ptab) should have the proper description
ok 231 - is_descendent_of(ctab, ptab) should have the proper diagnostics
ok 232 - is_descendent_of( ctab2, ptab ) should pass
ok 233 - is_descendent_of( ctab2, ptab ) should have the proper description
ok 234 - is_descendent_of( ctab2, ptab ) should have the proper diagnostics
ok 235 - is_descendent_of(nope, ptab) should fail
ok 236 - is_descendent_of(nope, ptab) should have the proper description
ok 237 - is_descendent_of(nope, ptab) should have the proper diagnostics
ok 238 - is_descendent_of(ctab2, nope) should fail
ok 239 - is_descendent_of(ctab2, nope) should have the proper description
ok 240 - is_descendent_of(ctab2, nope) should have the proper diagnostics
